src/OrleansConsulUtils/ConsulBasedMembershipTable.cs

```diff
@@ -0,0 +1,215 @@
+
+            return AssembleMembershipTableData(siloRegistration);
+        }
+
+        public async Task<MembershipTableData> ReadAll()
+        {
+            var deploymentKVAddresses = await _consulClient.KV.List(ConsulSiloRegistration.ParseDeploymentKVPrefix(_deploymentId));
+            if (deploymentKVAddresses.Response == null)
+            {
+                _logger.Verbose("Could not find any silo registrations for deployment {0}.", _deploymentId);
+                return new MembershipTableData(null);
+            }
+
+            var allSiloRegistrations = deploymentKVAddresses.Response.Select(siloKV => ConsulSiloRegistration.FromKVPair(_deploymentId, siloKV)).ToArray();
+
+            return AssembleMembershipTableData(allSiloRegistrations);
+        }
+
+        public async Task<Boolean> InsertRow(MembershipEntry entry, TableVersion tableVersion)
+        {
+            try
+            {
+                //Ensure the KV for the entry does not already exist
+                var value = await _consulClient.KV.Get(ConsulSiloRegistration.ParseDeploymentSiloKey(_deploymentId, entry.SiloAddress));
+                if (value.Response != null)
+                {
+                    _logger.Verbose("ConsulMembershipProvider could not insert a registration for silo {0}, because one already exists.", entry.SiloAddress);
+                    return false;
+                }
+
+                var consulSiloKVEntry = ConsulSiloRegistration.FromMembershipEntry(_deploymentId, entry);
+                await _consulClient.KV.Put(consulSiloKVEntry.ToKVPair());
+
+                return true;
+            }
+            catch (Exception ex)
+            {
+                _logger.Verbose("ConsulMembershipProvider failed to insert registration for silo {0}; {1}.", entry.SiloAddress, ex);
+                return false;
+            }
+        }
+
+        public async Task<Boolean> UpdateRow(MembershipEntry entry, String etag, TableVersion tableVersion)
+        {
+            //Update Silo Liveness
+            try
+            {
+                var siloRegistration = await GetConsulSiloRegistration(entry.SiloAddress);
+                if (siloRegistration == null)
+                {
+                    _logger.Verbose("ConsulMembershipProvider could not find a registration for silo {0}.", entry.SiloAddress);
+                    return false;
+                }
+
+                siloRegistration.Status = entry.Status;
+                siloRegistration.IAmAliveTime = entry.IAmAliveTime;
+                siloRegistration.SuspectingSilos.AddRange(entry.SuspectTimes.Select(silo => new SuspectingSilo { Id = silo.Item1.ToParsableString(), Time = silo.Item2 }).ToList());
+
+                var updateKV = siloRegistration.ToKVPair();
+
+                //If the KV.CAS() call returns false then the update failed
+                var tryUpdate = await _consulClient.KV.CAS(updateKV);
```

Comment:
I must have taken ignoring the TableVersion too far in that case, and now have it working as you have detailed here.   But there is one issue: The order of events during startup is Read > UpdateIAmAlive > UpdateRow.

The UpdateIAmAlive causing an unavoidable change to Consul's KV.ModifyIndex with no means of passing the new ModifyIndex/etag back to Orleans, using the etag Orleans then passes to UpdateRow for the CAS means it is guaranteed to fail on the first attempt, trigger a read and then succeed on the second call to UpdateRow.  Is this acceptable given your initial goal of keeping all the Row data in a single KV?